### PR TITLE
Generic shares_base module and specific s3_datasets_shares module - part 2 (db objects to shares_base)

### PR DESCRIPTION
**Pull request by dlpzx**
### Feature or Bugfix
- Refactoring

### Detail
As explained in the design for #1123 and #1283 we are trying to implement generic `datasets_base` and `shares_base` modules that can be used by any type of datasets and by any type of shareable object in a generic way. 

In this PR:
- Move shares state machines into its own file and move them to the `shares_base` module.
- Create new ShareObjectRepository and copy some generic methods to it
- Move db share objects to shares_base:
     - ShareObject has a field called datasetUri. For the scope of S3 and Redshift datasets we can leave it as dataseturi, but if we want to implement other kinds of sharing we should rethink it: we need to store the "approvers" of the share in some way. For the moment I am not going to go down that path until shares and s3 are uncoupled, then we can see how we would implement a complete generic sharing.
     - ShareObjectItem includes 3 fields related to the glue tables... we need to get rid of them in the backend code and look up the info with the itemType and itemUri. Left for part3 to keep the PR clean.

### Relates
- #1283 
- #1123 
- #955 

### Security
Please answer the questions below briefly where applicable, or write `N/A`. Based on
[OWASP 10](https://owasp.org/Top10/en/).

- Does this PR introduce or modify any input fields or queries - this includes
fetching data from storage outside the application (e.g. a database, an S3 bucket)?
  - Is the input sanitized?
  - What precautions are you taking before deserializing the data you consume?
  - Is injection prevented by parametrizing queries?
  - Have you ensured no `eval` or similar functions are used?
- Does this PR introduce any functionality or component that requires authorization?
  - How have you ensured it respects the existing AuthN/AuthZ mechanisms?
  - Are you logging failed auth attempts?
- Are you using or adding any cryptographic features?
  - Do you use a standard proven implementations?
  - Are the used keys controlled by the customer? Where are they stored?
- Are you introducing any new policies/roles/users?
  - Have you used the least-privilege principle? How?

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: backend/dataall/modules/shares_base/db/share_object_state_machines.py [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+                message=f'This transition is not possible, {prev_state} cannot go to {self._all_target_states}. '
+                f'If there is a sharing or revoking in progress wait until it is complete and try again.',
+            )
+        else:
+            return True
+
+    def get_transition_target(self, prev_state):
+        if self.validate_transition(prev_state):
+            for target_state, list_prev_states in self._transitions.items():
+                if prev_state in list_prev_states:
```
Comment (petrkalos) [inline backend/dataall/modules/shares_base/db/share_object_state_machines.py:38]: nit: `not in` and drop the else branch

File: backend/dataall/modules/shares_base/db/share_object_state_machines.py [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+from dataall.modules.shares_base.db.share_object_repositories import ShareObjectRepository
+
+logger = logging.getLogger(__name__)
+
+
+class Transition:
+    def __init__(self, name, transitions):
+        self._name = name
+        self._transitions = transitions
+        self._all_source_states = [*set([item for sublist in transitions.values() for item in sublist])]
```
Comment (petrkalos) [inline backend/dataall/modules/shares_base/db/share_object_state_machines.py:19]: nits
* `itertools.chain.from_iterable(transitions.values())` should flatten your values
*  why convert the set to list? I guess you want to dedup but why convert back to list?
*  by converting to set you assume that the values are hashable, I can see that are enum values so they must be but give it a thought

File: backend/dataall/modules/shares_base/db/share_object_state_machines.py [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+        self._all_target_states = [item for item in transitions.keys()]
```
Comment (petrkalos) [inline backend/dataall/modules/shares_base/db/share_object_state_machines.py:20]: nit: why is this needed? what's wrong with `self._all_target_states = transitions.keys()`?

File: backend/dataall/modules/shares_base/db/share_object_state_machines.py (hunk 1 above)
Comment (dlpzx) [reply backend/dataall/modules/shares_base/db/share_object_state_machines.py:38]: I just copy/pasted the code. I was very tempted to do changes on the implementation of the state machines, but I think we should do a review of them in a separate PR
